File: backend/app/grading.py
"""Deterministic move classification.

The LLM produces a hidden 0-100 interest score; classification is derived *server-side*
from the resulting swing (SPEC §3: "classification maps eval delta → label"), so thresholds
stay tunable and the label is never left to the model. Nothing persists the label — every move
table stores the numeric eval only, and this ramp derives the rank on read. The vocabulary
mirrors the frontend's `MoveClassKey` in `frontend/app/lib/game/types.ts`.
"""

from typing import Literal, NamedTuple

MoveClassKey = Literal["brilliant", "great", "good", "inaccuracy", "mistake", "blunder"]

# The single top rank — the move is already the strongest option, so it needs no "best line".
TOP_CLASS_KEY: MoveClassKey = "brilliant"

# The eval is on a 0-100 interest scale; "swing" is expressed in chess-style pawns = delta / 10
# (so a +24 interest jump reads as +2.4, matching the frontend's suggestion fixtures).
EVAL_PER_PAWN = 10.0

# Baseline interest a persona starts a fresh conversation with.
START_EVAL = 50.0


class Grade(NamedTuple):
    class_key: MoveClassKey
    glyph: str
    quality: int  # 0-100 weight used for running accuracy


# key -> (glyph, quality). Quality mirrors frontend MOVE_CLASSES.
_TABLE: dict[MoveClassKey, tuple[str, int]] = {
    "brilliant": ("!!", 100),
    "great": ("!", 88),
    "good": ("✓", 74),
    "inaccuracy": ("?!", 52),
    "mistake": ("?", 35),
    "blunder": ("??", 12),
}


def classify(swing: float) -> Grade:
    """Map a pawn-scale swing to a graded move. Monotonic ramp over SPEC §3 thresholds."""
    if swing >= 2.0:
        key: MoveClassKey = "brilliant"
    elif swing >= 1.0:
        key = "great"
    elif swing >= 0.2:
        key = "good"
    elif swing >= -1.0:
        key = "inaccuracy"
    elif swing >= -2.5:
        key = "mistake"
    else:
        key = "blunder"
    glyph, quality = _TABLE[key]
    return Grade(key, glyph, quality)


def swing_from_delta(eval_delta: float) -> float:
    return round(eval_delta / EVAL_PER_PAWN, 2)

File: frontend/packages/visuals/src/lib/grading.ts
import type { MoveClass, MoveClassKey } from "../types";

/**
 * Deterministic move grading — mirrors the server's `classify()` thresholds (backend
 * `app/grading.py`). Quality is stored as the numeric eval; the Brilliant…Blunder label is derived
 * on read, never persisted. The `--m-*` colors resolve against the shared theme tokens.
 */

/** Classification metadata, keyed by move class. */
export const MOVE_CLASSES: Record<MoveClassKey, MoveClass> = {
  checkmate_win: { key: "checkmate_win", glyph: "#", label: "Checkmate", color: "var(--m-checkmate-win)", quality: 100 },
  brilliant: { key: "brilliant", glyph: "!!", label: "Brilliant", color: "var(--m-brilliant)", quality: 96 },
  great: { key: "great", glyph: "!", label: "Great", color: "var(--m-great)", quality: 85 },
  good: { key: "good", glyph: "✓", label: "Good", color: "var(--m-good)", quality: 70 },
  inaccuracy: { key: "inaccuracy", glyph: "?!", label: "Inaccuracy", color: "var(--m-inaccuracy)", quality: 50 },
  mistake: { key: "mistake", glyph: "?", label: "Mistake", color: "var(--m-mistake)", quality: 30 },
  blunder: { key: "blunder", glyph: "??", label: "Blunder", color: "var(--m-blunder)", quality: 10 },
  checkmate_loss: { key: "checkmate_loss", glyph: "#", label: "Checkmate", color: "var(--m-checkmate-loss)", quality: 0 },
};

export function formatSwing(swing: number): string {
  return `${swing >= 0 ? "+" : ""}${swing.toFixed(1)}`;
}

/**
 * Derive a move's rank from its pawn-scale swing. Mirrors the server's `classify()` thresholds
 * (backend `app/grading.py`) — quality is stored as the numeric eval, the label is derived on
 * read. When `evalAfter` is supplied and sits on a mating square (≥100 / ≤0), the move is a
 * terminal checkmate regardless of the swing.
 */
export function classifySwing(swing: number, evalAfter?: number | null): MoveClassKey {
  if (evalAfter != null && evalAfter >= 100) return "checkmate_win";
  if (evalAfter != null && evalAfter <= 0) return "checkmate_loss";
  if (swing >= 2.5) return "brilliant";
  if (swing >= 1.2) return "great";
  if (swing >= 0.2) return "good";
  if (swing >= -0.8) return "inaccuracy";
  if (swing >= -2.0) return "mistake";
  return "blunder";
}

/** Same, from a raw eval delta (0–100 interest scale; swing = delta / 10). */
export function classifyEvalDelta(evalDelta: number | null, evalAfter?: number | null): MoveClassKey {
  return classifySwing((evalDelta ?? 0) / 10, evalAfter);
}
